File: day01/pascal/part2.pas
{$mode objfpc}{$H+}{$J-}

program Part2;

uses
  ArrayUtils, JabbaTypes, JabbaUtils, sysutils;


function CountElements(const arr: array of Integer; const value: Integer): Integer;
var
  i: Integer;
begin
  Result := 0;
  for i := Low(arr) to High(arr) do
    if arr[i] = value then
      Result += 1;
end;


procedure Main();
var
  fname, line: String;
  lines, parts: TStringArray;
  col1, col2: TIntArray;
  idx: Integer = 0;
  total: Integer = 0;
  value: Integer;
begin
  // fname := 'example.txt';
  fname := 'input.txt';
  lines := Readlines(fname);
  col1 := [];
  col2 := [];
  for line in lines do
    begin
      parts := PySplit(line);
      SetLength(col1, Length(col1) + 1);
      SetLength(col2, Length(col2) + 1);
      col1[idx] := StrToInt(parts[0]);
      col2[idx] := StrToInt(parts[1]);
      idx += 1;
    end;
  for idx := Low(col1) to High(col1) do
    begin
      value := col1[idx];
      total += value * CountElements(col2, value);
    end;
  WriteLn(total);
end;

//---------------------------------------------------------------------------

begin
  WriteLn('---');
  Main();
end.

File: day01/python/part2.py
#!/usr/bin/env python3

"""
3   4
4   3
2   5
1   3
3   9
3   3

d:
    4: 1
    3: 3
    5: 1
    9: 1

"""

import helper


def main() -> None:
    # lines: list[str] = helper.read_lines("example.txt")
    lines: list[str] = helper.read_lines("input.txt")

    first: list[int] = []
    second: list[int] = []

    for line in lines:
        parts = line.split()
        first.append(int(parts[0]))
        second.append(int(parts[1]))
    #

    result = 0
    for n in first:
        result += n * second.count(n)
    #
    print(result)


##############################################################################

if __name__ == "__main__":
    main()
